keps/sig-multicluster/2149-clusterid/README.md

```diff
@@ -0,0 +1,756 @@
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### ClusterSet membership
+
+I have some set of clusters working together and need a way to uniquely identify them within the system that I use to track membership.
+
+#### Multi-Cluster Services
+
+I have a headless multi-cluster service deployed across clusters in my ClusterSet with similarly named pods in each cluster. I need a way to disambiguate each backend pod via DNS.
+
+```
+<<[UNRESOLVED]>>
+Examples of DNS using cluster ID a la `<hostname>.<clusterID>.<svc>.<ns>.svc.clusterset.local`
+<<[/UNRESOLVED]>>
+```
+
+#### Diagnostics
+
+Clusters within my ClusterSet send logs/metrics to a common monitoring solution and I need to be able to identify the cluster from which a given set of events originated.
+
+
+#### Joining or moving between ClusterSets
+
+I want the ability to add a previously-isolated cluster to a ClusterSet, or to move a cluster from one ClusterSet to another and be aware of this change.
+
+### `ClusterClaim` CRD
+```
+<<[UNRESOLVED]>>
+The actual name of the CRD is not finalized and is provisionally titled `ClusterClaim` for the remainder of this document.
+<<[/UNRESOLVED]>>
+```
+
+The `ClusterClaim` resource provides a way to store identification related, cluster scoped information for multi-cluster tools while creating flexibility for implementations. A cluster may have multiple `ClusterClaim`s, each holding a different identification related value. Each claim contains the following information:
+
+*   **Name** - a well known or custom name to identify the claim.
+*   **Value** - a claim-dependent string.
+
+
+### Well known claims
+
+The requirements below use the keywords **must, should,** and **may** purposefully in accordance with [RFC-2119](https://tools.ietf.org/html/rfc2119).
+
+
+#### Claim: `id.k8s.io`
+
+Contains a unique identifier for the containing cluster.
+
+
+##### Uniqueness
+
+*   The identifier **must** be unique within the ClusterSet to which its cluster belongs for the duration of the cluster’s membership.
+*   The identifier **should** be unique beyond the ClusterSet within the scope of expected use.
+*   The identifier **may** be globally unique beyond the scope of its ClusterSet.
+*   The identifier **may** be unique beyond the span of its cluster’s membership and lifetime.
+
+
+##### Lifespan
+
+*   The identifier **must** exist and be immutable for the duration of a cluster’s membership in a ClusterSet, and as long as a `clusterset.k8s.io` claim referring to that cluster in that ClusterSet exists.
+*   The identifier **must** exist for the lifespan of a cluster.
+*   The identifier **may** be immutable for the lifespan of a cluster.
```

Comment:
Can this become "should" ?  By doing so, we'd be encouraging tools like kubeadm to make this an important decision point, and likely emphasize that ID is more "automation oriented" than human oriented.

Reply:
I think **should** is better here, but we probably also want to discuss how that applies to clusters that were originally isolated and only acquired an id when joining a clusterset

Reply:
Changed to should in 71ff342, and added unresolved discussion point on the scenario of previously isolated clusters getting IDs in 4c1d73d